utils/server_auth.go

```diff
@@ -0,0 +1,342 @@
+ * may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+package utils
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+	"github.com/pelicanplatform/pelican/config"
+	"github.com/pelicanplatform/pelican/director"
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+type (
+	TokenSource int
+	TokenIssuer int
+	AuthOption  struct {
+		Sources []TokenSource
+		Issuers []TokenIssuer
+		Scopes  []string
+	}
+	AuthChecker interface {
+		FederationCheck(ctx *gin.Context, token string, scopes []string) error
+		DirectorCheck(ctx *gin.Context, token string, scopes []string) error
+		IssuerCheck(ctx *gin.Context, token string, scopes []string) error
+	}
+	AuthCheckImpl struct{}
+)
+
+const (
+	Header TokenSource = iota // "Authorization" header
+	Cookie                    // "login" cookie
+	Authz                     // "authz" query parameter
+)
+
+const (
+	Federation TokenIssuer = iota
+	Director
+	Issuer
+)
+
+var (
+	federationJWK *jwk.Cache
+	authChecker   AuthChecker
+)
+
+func init() {
+	authChecker = &AuthCheckImpl{}
+}
+
+// Return if desiredScopes contains the tokenScope and it's case-insensitive
+func scopeContains(tokenScope string, desiredScopes []string) bool {
+	for _, sc := range desiredScopes {
+		if strings.EqualFold(sc, tokenScope) {
+			return true
+		}
+	}
+	return false
+}
+
+// Creates a validator that checks if a token's scope matches the given scope: matchScope.
+// Will pass the check if no "anyScopes".
+// Will pass the check if one token scope matches ANY item in "anyScopes"
+func createScopeValidator(anyScopes []string) jwt.ValidatorFunc {
+
+	return jwt.ValidatorFunc(func(_ context.Context, tok jwt.Token) jwt.ValidationError {
+		// If no scope is present, always return true
+		if len(anyScopes) == 0 {
+			return nil
+		}
+		scope_any, present := tok.Get("scope")
+		if !present {
+			return jwt.NewValidationError(errors.New("No scope is present; required for authorization"))
+		}
+		scope, ok := scope_any.(string)
+		if !ok {
+			return jwt.NewValidationError(errors.New("scope claim in token is not string-valued"))
+		}
+
+		for _, tokenScope := range strings.Split(scope, " ") {
+			// As long as there's one scope in the token that matches the pool of desriedScopes
+			// we say it's valid
```

Comment:
I wonder about the semantics of this. Let's say I want to know if a token gives me permission to read and write some resource, and let's assume that the token does actually contain both scopes needed for those actions. To actually verify that using this validator, wouldn't I need to validate it twice, once for each scope?

Is there a drawback to requiring that every scope passed in the `anyScopes` is matched in the token?

Reply:
You are right, this is also the part that I don't have a lot of confidence when implementing it. The `anyToken` check was designed for the case where a single API can accept multiple valid scopes for the same resource as long as there is one exists in the token (so that we don't need to check each of the scopes). This is no longer the case with the scope refactoring we made as part of this PR so I guess it's less useful. What I will to is to add the additional parameter (like `checkAllScopes`) to the function and enable the feature you described, which was supposed to be done in the previous ticket but I just didn't get enough effort in making that happen (my bad)